File: Data Structures & Algorithms/combination-target-sum-ii/submission-3.java
class Solution {

    public void solve(int index, int target, int[] nums, 
        List<Integer> tmp, List<List<Integer>> ans, int sum){

            if(target == sum){
                ans.add(new ArrayList<>(tmp));
                return ;
            }     

            for(int i=index; i<nums.length; i++){
                if(i>index && nums[i] == nums[i-1])
                    continue;

                if(nums[i] > target - sum)
                    break;

                tmp.add(nums[i]);
                solve(i+1, target, nums, tmp, ans, sum+nums[i]);
                tmp.remove(tmp.size()-1);
            }
    } 

    public List<List<Integer>> combinationSum2(int[] nums, int target) {
        List<Integer> tmp = new ArrayList<>();
        List<List<Integer>> ans = new ArrayList<>();
        Arrays.sort(nums);
        solve(0, target, nums, tmp, ans, 0);


        return ans;
    }
}

File: Data Structures & Algorithms/search-2d-matrix/submission-3.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int l=0, r=matrix.size()*matrix[0].size()-1;

        while(l<=r){
            int mid = l + (r-l)/2;
            int m = mid/matrix[0].size(), n = mid%matrix[0].size();

            if(matrix[m][n] == target)
                return true;
            else if(matrix[m][n] > target){
                r=mid-1;
            }
            else
                l=mid+1;
        }
        return false;
    }
};
